### ISSUE TITLE
Publish pyshaper to PyPI

### ISSUE THREAD

**Issue by matthewfeickert** [rikab/SHAPER#7]
Okay, so here's my suggestions of steps for deploying to TestPyPI and PyPI manually

## Get accounts

1. [Register for account on TestPyPI](https://test.pypi.org/account/register/)
2. [Register for account on PyPI](https://pypi.org/account/register/)

## Setup virtual environment for deployment

1. Activate your virtual environment (or create a new one if needed)
2. Update `pip`, `setuptools`, `wheel`

```
python -m pip install --upgrade pip setuptools wheel
```

3. Install `build` and `twine`

```
python -m pip install --upgrade build twine
```

## Deploy tag `v1.0.0`

There already exists the tag/release on GitHub [`v1.0.0`](https://github.com/rikab/SHAPER/releases/tag/v1.0.0) which you just made. However, that isn't the current `HEAD` of the `main` branch, so to get the correct version number in the distributions we can rewind to the commit with the [`v1.0.0`](https://github.com/rikab/SHAPER/releases/tag/v1.0.0) tag and deploy _that_.

1. For simplicity, clone the project fresh from GitHub

```
cd /tmp
git clone git@github.com:rikab/SHAPER.git
cd SHAPER
```

2. Reset to the state of the `v1.0.0` tag

```
git reset --hard v1.0.0
```

4. Build a sdist and wheel distribution with `build`

```console
$ python -m build .
...
Successfully built pyshaper-1.0.0.tar.gz and pyshaper-1.0.0-py3-none-any.whl
```

5. Check the state of the distributions with `twine`

```console
$ twine check dist/*
Checking dist/pyshaper-1.0.0-py3-none-any.whl: PASSED
Checking dist/pyshaper-1.0.0.tar.gz: PASSED
```

6. List the contents of the sdist (which is a tarball) and the wheel (which is actually a zipfile) to manually inspect them

```
python -m tarfile --list dist/pyshaper-*.tar.gz
```
```
python -m zipfile --list dist/pyshaper-*.whl
```

The wheel will be a subset of the sdist (_source_ distribution) that just contains the actual package files.

7. Upload the distributions to TestPyPI (this will require your TestPyPI username and password from before)

```
twine upload --repository testpypi dist/*
```

9. Check that things look good on TestPyPI: https://test.pypi.org/project/pyshaper/ :+1:

Assuming they do...

10. Upload the distributions to PyPI (this will require your PyPI username and password from before)

```
twine upload dist/*
```

Congratulations! You have just released a project to the Python Package Index! https://pypi.org/project/pyshaper/ :tada:

## Register API tokens for your project

Now that you have gone through the very manual first time procedure you can setup API tokens to make this easier in the future.

### TestPyPI

1. Login to [TestPyPI](https://test.pypi.org/manage/account/token/)
2. Create an [API token](https://pypi.org/help/#apitoken) for your project.
   * [Account settings -> API tokens](https://test.pypi.org/manage/account/token/)
      - **Token name:** pyshaper deployment (or something similarly descriptive)
      - **Scope:** Project: pyshaper
   * **Save the token value to your password manager and note this is for TestPyPI**

### PyPI

1. Login to [PyPI](https://test.pypi.org/manage/account/token/)
2. Create an [API token](https://pypi.org/help/#apitoken) for your project.
   * [Account settings -> API tokens](https://test.pypi.org/manage/account/token/)
      - **Token name:** pyshaper deployment (or something similarly descriptive)
      - **Scope:** Project: pyshaper
   * **Save the token value to your password manager and note this is for PyPI**

## General deployment strategy in the future

I have a totally CI based workflow that I can share in the future that allows for deploying to TestPyPI and PyPI with [just a few button clicks](https://www.youtube.com/watch?v=tsAWDumpcW8). Here's a manual workflow though.

1. From the commit that you want to release from (probably the latest commit on GitHub (`origin/main`)) bump the version information in your repository to the SemVer version you want and also create a corresponding tag at that commit.

You can use a tool like [`tbump`](https://github.com/your-tools/tbump) to do this.

2. Push that commit with the new versions and tag to GitHub

```
git push origin HEAD --tags
```

3. Build a sdist and wheel distribution with `build`

```
rm -rf build
python -m build .
```

4. Check the state of the distributions with `twine`

```
twine check dist/*
```

5. List the contents of the sdist (which is a tarball) and the wheel (which is actually a zipfile) to manually inspect them

```
python -m tarfile --list dist/pyshaper-*.tar.gz
```
```
python -m zipfile --list dist/pyshaper-*.whl
```

6. Upload the distributions to TestPyPI (this will require your TestPyPI username and API token)

```
twine upload --repository testpypi dist/*
```

7. Check that things look good on TestPyPI: https://test.pypi.org/project/pyshaper/ :+1:

Assuming they do...

8. Upload the distributions to PyPI (this will require your PyPI username and API token)

```
twine upload dist/*
```

9. Create a [GitHub release](https://github.com/rikab/SHAPER/releases) from the same tag you just deployed to PyPI to have corresponding releases across the platforms.

_Originally posted by @matthewfeickert in https://github.com/rikab/SHAPER/pull/5#discussion_r1122409413_

**Comment by matthewfeickert** [comment]
I made this its own Issue so that it would be easier to link to in the future as a point of reference. (I should make a generic version of this as a GitHub Gist at some point.)

**Comment by rikab** [comment]
Ah, so I made tiny changes to the readme/citation file after the 1.0.0 tag, which in hindsight I should have done in the opposite order. Is it best to just undo these? Or call it 1.0.1?

**Comment by matthewfeickert** [comment]
I'd still deploy tag `v1.0.0` like I describe above with a reset of history. You can always make patch releases as you'd like later.

**Comment by rikab** [comment]
I think I've gotten PyPi and pip to work! I'm a bit unclear about this last step however. Do you mean that I should create a new release (1.0.1)? Is it different in any way from the 1.0.0 release?

**Comment by matthewfeickert** [comment]
> I think I've gotten PyPI and pip to work!

Indeed! Congrats! :tada: 

```console
$ docker run --rm -ti python:3.10 /bin/bash
root@d48d47bac772:/# python -m venv venv && . venv/bin/activate
(venv) root@d48d47bac772:/# python -m pip --quiet install --upgrade pip setuptools wheel
(venv) root@d48d47bac772:/# python -m pip --quiet install numpy torch
(venv) root@d48d47bac772:/# python -m pip --quiet install --upgrade pyshaper
(venv) root@d48d47bac772:/# python -m pip show pyshaper
Name: pyshaper
Version: 1.0.0
Summary: framework for defining, building, and evaluating generalized shape observables for collider physics
Home-page: 
Author: 
Author-email: Rikab Gambhir <rikab@mit.edu>
License: MIT
Location: /venv/lib/python3.10/site-packages
Requires: geomloss, matplotlib, numpy, pyjet, scipy, torch
Required-by: 
(venv) root@d48d47bac772:/# python -c 'import pyshaper; print(pyshaper)'
<module 'pyshaper' from '/venv/lib/python3.10/site-packages/pyshaper/__init__.py'>
(venv) root@d48d47bac772:/# 
```

This is mission accomplished for this Issue, so I'm going to close it but we can continue discussing here until all your questions are resolved. :+1: 

> I'm a bit unclear about this last step however.

Hm, which step specifically? The "General deployment strategy in the future" stuff? 

> Do you mean that I should create a new release (1.0.1)? Is it different in any way from the 1.0.0 release?

Anytime that you want to make a new release of a distribution to PyPI (following the steps in "General deployment strategy in the future") you should make a new tag and a new GitHub release to accompany it.

The thing to be careful about, and why I mention tools like [`tbump`](https://github.com/your-tools/tbump) in the above is you'll note that you now have version numbers in the documentation

```console
$ git grep "1\.0\.0"
CITATION.cff:version: v1.0.0
README.md:# SHAPER (v1.0.0)
README.md:      title = "{pyshaper: v1.0.0}",
README.md:      version = {1.0.0},
README.md:      note = {https://github.com/rikab/SHAPER/releases/tag/v1.0.0}
README.md:* v1.0.0: 24 February 2023. Official public release.
examples/example.ipynb:    "* `lr` (default is .01) is the learning rate for the ADAM optimizer, used to optimize shape parameters. As of `SHAPER v1.0.0`, ADAM is the only supported optimizer.\n",
```

so whenever you make a new release (be that a patch release `v1.0.1` or a minor release `v1.1.0` or whatever) you'll want to make sure **in the commit that you make that will get tagged with the new release version** you also update these `v1.0.0`s to the new version tag.

Explicitly this looks like the following ones would need to get bumped:

https://github.com/rikab/SHAPER/blob/e099132e985500749f9da586bbc8d2c3a7d49a9a/CITATION.cff#L32

https://github.com/rikab/SHAPER/blob/e099132e985500749f9da586bbc8d2c3a7d49a9a/README.md?plain=1#L1

https://github.com/rikab/SHAPER/blob/e099132e985500749f9da586bbc8d2c3a7d49a9a/README.md?plain=1#L60

https://github.com/rikab/SHAPER/blob/e099132e985500749f9da586bbc8d2c3a7d49a9a/README.md?plain=1#L61

https://github.com/rikab/SHAPER/blob/e099132e985500749f9da586bbc8d2c3a7d49a9a/README.md?plain=1#L64

If you need an example of how this might look you can checkout [`pyhf`'s `tbump.toml`](https://github.com/scikit-hep/pyhf/blob/a4ec176d07a7484c2e4829ed69f08ad3bc3e6af6/tbump.toml) and you can feel free to ask me if things are unclear (just ping me on an Issue on this repo or something).
